jsonmessage: add and use TimeNano for events

vbatts opened the pull request:
The events logging uses nano formatting, but only provides a Unix()
time, therefor ordering may get lost in the output. Example:

```
2015-09-15T14:18:51.000000000-04:00 ee46febd64ac629f7de9cd8bf58582e6f263d97ff46896adc5b508db804682da: (from busybox) resize
2015-09-15T14:18:51.000000000-04:00 a78c9149b1c0474502a117efaa814541926c2ae6ec3c76607e1c931b84c3a44b: (from busybox) resize
```

By having a field just for Nano time, when set, the marshalling back to
`time.Unix(sec int64, nsec int64)` has zeros exactly where it needs to.
This does not break any existing use of jsonmessage.JSONMessage, but now
allows for use of `UnixNano()` and get event formatting that has
distinguishable order. Example:

```
2015-09-15T15:37:23.810295632-04:00 6adcf8ed9f5f5ec059a915466cd1cde86a18b4a085fc3af405e9cc9fecbbbbaf: (from busybox) resize
2015-09-15T15:37:23.810412202-04:00 6b7c5bfdc3f902096f5a91e628f21bd4b56e32590c5b4b97044aafc005ddcb0d: (from busybox) resize
```

Signed-off-by: Vincent Batts vbatts@redhat.com

vbatts committed:
events/jsonmessage: add and prefer TimeNano for events

This way provide both Time and TimeNano in the event. For the display of
the JSONMessage, use either, but prefer TimeNano Proving only TimeNano
would break Subscribers that are using the `Time` field, so both are set
for backwards compatibility.

The events logging uses nano formatting, but only provides a Unix()
time, therefor ordering may get lost in the output. Example:
```
2015-09-15T14:18:51.000000000-04:00 ee46febd64ac629f7de9cd8bf58582e6f263d97ff46896adc5b508db804682da: (from busybox) resize
2015-09-15T14:18:51.000000000-04:00 a78c9149b1c0474502a117efaa814541926c2ae6ec3c76607e1c931b84c3a44b: (from busybox) resize
```

By having a field just for Nano time, when set, the marshalling back to
`time.Unix(sec int64, nsec int64)` has zeros exactly where it needs to.
This does not break any existing use of jsonmessage.JSONMessage, but now
allows for use of `UnixNano()` and get event formatting that has
distinguishable order. Example:
```
2015-09-15T15:37:23.810295632-04:00 6adcf8ed9f5f5ec059a915466cd1cde86a18b4a085fc3af405e9cc9fecbbbbaf: (from busybox) resize
2015-09-15T15:37:23.810412202-04:00 6b7c5bfdc3f902096f5a91e628f21bd4b56e32590c5b4b97044aafc005ddcb0d: (from busybox) resize
```

Including tests for TimeNano and updated event API reference doc.

Signed-off-by: Vincent Batts <vbatts@redhat.com>

vbatts commented:
discovered while testing for https://github.com/docker/docker/pull/16316

Ping @LK4D4

duglin commented:
@vbatts do we want to add a testcase to make sure we don't drop the extra number by mistake later on?

vbatts commented:
@duglin i reckon ... ;-)

vbatts commented:
@duglin done. PTAL

duglin commented:
@vbatts thanks!

LK4D4 commented on pkg/jsonmessage/jsonmessage.go:
```diff
@@ -99,6 +99,7 @@ type JSONMessage struct {
 	ID              string        `json:"id,omitempty"`
 	From            string        `json:"from,omitempty"`
 	Time            int64         `json:"time,omitempty"`
+	TimeNano        int64         `json:"timeNano,omitempty"`
```
Uh-oh :)

vbatts replied:
What?

LK4D4 replied:
New field, looks sorta weird.

vbatts replied:
Nah. It's straight forward and logical. Trying to overload the existing
field would be a mess.
On Sep 15, 2015 5:43 PM, "Alexander Morozov" <notifications@github.com>
wrote:

> In pkg/jsonmessage/jsonmessage.go
> <https://github.com/docker/docker/pull/16317#discussion_r39569479>:
>
> > @@ -99,6 +99,7 @@ type JSONMessage struct {
> >  	ID              string        `json:"id,omitempty"`
> >  	From            string        `json:"from,omitempty"`
> >  	Time            int64         `json:"time,omitempty"`
> > +	TimeNano        int64         `json:"timeNano,omitempty"`
>
> New field, looks sorta weird.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16317/files#r39569479>.
>

calavera replied:
but the existent field is not set anymore, is it? I don't know how I feel about a field that's never set.

vbatts replied:
It's present. And I thought we tried to not break API for ./pkg/?
On Sep 15, 2015 19:23, "David Calavera" <notifications@github.com> wrote:

> In pkg/jsonmessage/jsonmessage.go
> <https://github.com/docker/docker/pull/16317#discussion_r39578388>:
>
> > @@ -99,6 +99,7 @@ type JSONMessage struct {
> >  	ID              string        `json:"id,omitempty"`
> >  	From            string        `json:"from,omitempty"`
> >  	Time            int64         `json:"time,omitempty"`
> > +	TimeNano        int64         `json:"timeNano,omitempty"`
>
> but the existent field is not set anymore, is it? I don't know how I feel
> about a field that's never set.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16317/files#r39578388>.
>

vbatts commented:
fixed API break for event subscribers and still providing Nano accuracy.

LK4D4 commented:
Requires API docs update :)

LK4D4 commented:
Code looks correct for me, though

vbatts commented:
@LK4D4 added remote API reference update

LK4D4 commented:
@vbatts You need also add this to `docker_remote_api.md` in `### v1.21 API changes`

vbatts commented:
@LK4D4  updated

cpuguy83 commented:
Needs a rebase already :(

vbatts commented:
rebased

calavera commented:
LGTM. Do you mind squashing those little commits into one? :pray:

cpuguy83 commented:
Also _another_ rebase.

vbatts commented:
rebased and squashed. PTAL.

On Wed, Sep 16, 2015 at 5:10 PM, Brian Goff <notifications@github.com>
wrote:

> Also *another* rebase.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16317#issuecomment-140895916>.
>

cpuguy83 commented:
LGTM

cpuguy83 commented:
ping @thaJeztah

thaJeztah commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -2006,10 +2006,10 @@ and Docker images report:
+    {"status":"pull","id":"busybox:latest","time":1442421700,"timeNano":1442421700598988358}
+    {"status":"create","id":"5745704abe9caa57898ed7c1e2f85b73b00e2b128a2d681c74bf4c46d00a717d","from":"busybox","time":1442421716,"timeNano":1442421716853979870}
+    {"status":"attach","id":"5745704abe9caa57898ed7c1e2f85b73b00e2b128a2d681c74bf4c46d00a717d","from":"busybox","time":1442421716,"timeNano":1442421716894759198}
+    {"status":"start","id":"5745704abe9caa57898ed7c1e2f85b73b00e2b128a2d681c74bf4c46d00a717d","from":"busybox","time":1442421716,"timeNano":1442421716983607193}
```
Hm, bit worried that the useful information will now be hidden behind a scrollbar.

Thinking how to best solve this;
"cheat", and use a short ID in the output (did this ever return a short-ID as in the old output?), or
reformat the JSON;

```JSON
{
    "status": "create",
    "id": "5745704abe9caa57898ed7c1e2f85b73b00e2b128a2d681c74bf4c46d00a717d",
    "from": "busybox",
    "time": 1442421716,
    "timeNano": 1442421716853979870
}
```

If we *do* reformat the JSON, we should add a note (similar to [what's here](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.21.md#get-an-archive-of-a-filesystem-resource-in-a-container), i.e. " (whitespace added for readability)"

thaJeztah commented:
Thanks @vbatts, I added a note, but apart from that, LGTM

ping @moxiegirl

LK4D4 commented:
but it needs rebase again :/

cpuguy83 commented:
holy :hankey:

thaJeztah commented:
Time between rebases is now measured in nanoseconds!

vbatts commented:
lol.
rebased and shortened the ID in the doc example

thaJeztah commented:
Thanks, LGTM

ping @moxiegirl @SvenDowideit

thaJeztah commented:
janky failed, dunno if it's related;
```
FAIL: dockerCmd_utils_test.go:161: DockerCmdSuite.TestDockerCmdWithTimeout

dockerCmd_utils_test.go:205:
    c.Assert(out, check.Equals, cmd.expectedOut, check.Commentf("Expected output %q for arguments %v, got %q", cmd.expectedOut, cmd.args, out))
... obtained string = "too long, should be killed"
... expected string = ""
... Expected output "" for arguments [a command that times out], got "too long, should be killed"

cpuguy83 commented:
It's not related.

moxiegirl commented:
LGTM @vbatts

LK4D4 commented:
@vbatts Last rebase before merge, pls :)

vbatts commented:
rebased. PTAL.

cpuguy83 commented:
All good, merging.
